### PR TITLE
feat: optimize ledger to increase TPS

### LINKED ISSUE #671: Ledger supports custom index
### Description of the issue

- add table schema to support custom index
- refactor table id
- refactor ledger generic to support cache

### Issue-Type
- [ ] bug report
- [x] feature request
- [ ] Documentation improvement

### LINKED ISSUE #773: add cache layer to block process
### Description of the issue
Describe the issue you are experiencing.

_Please put an `x` against the checkboxes._  

### Issue-Type
- [ ] bug report
- [ ] feature request
- [ ] Documentation improvement

### Checklist
- [ ] Running latest version of code.
- [ ] This issue has not been reported earlier.

### Your environment
* OS
* Go version
* Release tag/commit of the code

### Expected behavior
What should happen?

### Actual behavior
What is actually happening?

### Steps to reproduce
1. Provide detailed steps
2. how to reproduce this issue

### Any extra info ( for eg. code snippet to reproduce, logs, etc. )
If necessary, provide some extra information like code-snippets or error-logs.

### LINKED ISSUE #781: optimize ledger to increase tps
- memory cache #773
   - [x] implement cache and sqlite index interface #798
   - [x] cache management, write-cache switch/read-cache update
   - [ ] dump to badger and sqlite
   - [ ] refactor ledger generic to support cache
- badger
   - [x] move store interface to common folder
- relation db #671
   - [ ] table schema interface of each struct
   - [ ] implement table query
   - [ ] migrate sqlite service into ledger service #783
   - [ ] support pq/mysql etc.
- log file
   - [ ] node exception exit handling
- ledger
  - [x] refactor ledger interface
  - [ ] refactor block check/process/rollback
  - [ ] synchronize block / unconfirmed block 
  - [ ] refactor ledger migration
  - [ ] ledger dump #811
  - [ ] testcase

### LINKED ISSUE #783: migrate sqlite service into ledger service
### Description of the issue

- support pq/mysql etc.
- migration sqlite service into ledger service

### Issue-Type
- [ ] bug report
- [x] feature request
- [ ] Documentation improvement

### LINKED ISSUE #927: verify contract block fail when node synchronization
### Description of the issue
Describe the issue you are experiencing.

_Please put an `x` against the check boxes._  

### Issue-Type
- [x] bug report
- [ ] question
- [ ] Documentation improvement

### PR DESCRIPTION
**Pull request by zengchen221**
### Proposed changes in this pull request

optimize ledger to increase tps, fix #781 
 - add cache layer to block process, fix  #773 
 - refactor ledger interface
 - ledger supports custom index, fix  #671 
 - migrate sqlite service into ledger service
 - fix #783 
 - fix #927 

### Type
- [x] Bug fix: (Link to the issue #{issue No.})
- [x] Feature (Non-breaking change)
- [ ] Feature (Breaking change)
- [ ] Documentation Improvement

### Checklist
Write a small comment explaining if its `N/A` (not applicable)

- [x] Read the [CONTRIBUTION](https://github.com/qlcchain/go-qlc/blob/master/CONTRIBUTING.md).
- [x] All the tests are passing after the introduction of new changes.
- [ ] Added tests respective to the part of code I have written.
- [ ] Added proper documentation where ever applicable.
- [x] Code has been written according to [Golang-Style-Guide](https://github.com/qlcchain/go-qlc/blob/master/CONTRIBUTING.md#code-standard)

### Extra information
Any extra information related to this pull request.

### CODE REVIEW

File: common/storage/storage.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+			src = v[:]
+		case types.Address:
+			src = v[:]
+		case *types.PendingKey:
+			pk := part.(*types.PendingKey)
+			var err error
+			src, err = pk.Serialize()
+			if err != nil {
+				return nil, fmt.Errorf("pending key serialize: %s", err)
+			}
```
Comment (gythialy) [inline common/storage/storage.go]: define `Serializer`interface to make it more generic?

File: common/types/block_state.go [hunk 2]
```diff
@@ -233,6 +234,44 @@ func (bs *StateBlockList) Deserialize(text []byte) error {
...
+			hash %s,
+			type %s,
+			address %s,
+			timestamp %s
+		)`,
+		b.TableName(),
+		ConvertSchemaType(b.GetHash()),
+		ConvertSchemaType(b.Type),
+		ConvertSchemaType(b.Address),
+		ConvertSchemaType(b.Timestamp))
```
Comment (gythialy) [inline common/types/block_state.go]: How to deal with different databases?

File: common/types/schma.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
+package types
+
+type Serialize interface {
```
Comment (gythialy) [inline common/types/schma.go]: file name `schma`->`schema`

Comment (zengchen221) [reply common/types/schma.go]: fixed

File: common/types/block_state.go (hunk 2 above)
Comment (zengchen221) [reply common/types/block_state.go]: i add a parameter to distinguish database， i will test mysql later

File: common/storage/storage.go (hunk 1 above)
Comment (zengchen221) [reply common/storage/storage.go]: done, `Serializer` interface  is defined in common/types folder

**Pull request merged**